scripts/core/editor3/components/spellchecker/Spellchecker.jsx

```diff
@@ -0,0 +1,229 @@
+/**
+ * @ngdoc React
+ * @module superdesk.core.editor3
+ * @name Spellchecker
+ * @param {Function} getEditorState
+ * @param {Function} setEditorState
+ * @param {Object} spellchecker
+ * @param {Object} editorRect
+ * @description check the text and shows the inline spellchecker context menu
+ */
+
+import React from 'react';
+import {EditorState, Modifier} from 'draft-js';
+
+export class Spellchecker extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.addWord = this.addWord.bind(this);
+        this.replaceWord = this.replaceWord.bind(this);
+
+        this.isVisible = this.isVisible.bind(this);
+    }
+
+    /**
+      * @ngdoc method
+      * @name Spellchecker#getDecorator
+      * @param {Function} getEditorState
+      * @param {Function} setEditorState
+      * @param {Object} spellchecker
+      * @return {Object} returns the decorator
+      * @description Decorator for spellchecker, will highlight the words with errors
+      */
+    static getDecorators(getEditorState, setEditorState, spellchecker) {
+        return [{
+            strategy: spellcheckStrategy(spellchecker.isCorrectWord),
+            component: Spellchecker.spellcheckSpan,
+            props: {
+                getEditorState: getEditorState,
+                setEditorState: setEditorState,
+                spellchecker: spellchecker,
+                showContextMenu: Spellchecker.showContextMenu
+            }
+        }];
+    }
+
+    /**
+      * @ngdoc method
+      * @name Spellchecker#spellcheckSpan
+      * @return {Object} return the component used to highlight errors
+      * @description Component to highlight errors
+      */
+    static spellcheckSpan(props) {
+        const style = styleMap.ERROR;
+        const onContextMenu = (e) => props.showContextMenu(e, props);
+
+        return <span style={style} onContextMenu={onContextMenu}>{props.children}</span>;
+    }
+
+    /**
+      * @ngdoc method
+      * @name Spellchecker#showContextMenu
+      * @param {Object} e on context menu event
+      * @param {Object} the properties of the component that generates the event
+      * @description Show the contextual menu, init the suggestions list
+      */
+    static showContextMenu(e, props) {
+        const left = e.clientX;
+        const top = e.clientY;
+        var element = props.children[0];
+
+        e.preventDefault();
+        props.spellchecker.suggest(element.props.text)
+        .then((suggestions) => {
+            var {editorState} = props.getEditorState();
+
+            props.setEditorState({
```

Comment:
This is not related to decorator imo, but an action triggered on spellcheck span component, those 2 could be defined outside of `Spellchecker` which is not a component, but rather a service that gets words list and suggestions.
When you pick a suggestions it should change editor state, but not when you render the dropdown.